File: src/main/java/edu/cornell/cs/apl/viaduct/imp/informationflow/LabelConstant.java
package edu.cornell.cs.apl.viaduct.imp.informationflow;

import edu.cornell.cs.apl.viaduct.security.FreeDistributiveLattice;
import edu.cornell.cs.apl.viaduct.security.Label;
import edu.cornell.cs.apl.viaduct.security.Principal;
import edu.cornell.cs.apl.viaduct.security.solver.ConstantTerm;

/** A statically known constant. */
final class LabelConstant extends LabelTerm {
  private final ConstantTerm<FreeDistributiveLattice<Principal>> confidentiality;
  private final ConstantTerm<FreeDistributiveLattice<Principal>> integrity;
  private final Label value;

  LabelConstant(
      ConstantTerm<FreeDistributiveLattice<Principal>> confidentiality,
      ConstantTerm<FreeDistributiveLattice<Principal>> integrity)
  {
    this.confidentiality = confidentiality;
    this.integrity = integrity;
    this.value = new Label(confidentiality.getValue(), integrity.getValue());
  }

  @Override
  public Label getValue() {
    return value;
  }

  @Override
  ConstantTerm<FreeDistributiveLattice<Principal>> getConfidentiality() {
    return this.confidentiality;
  }

  @Override
  ConstantTerm<FreeDistributiveLattice<Principal>> getIntegrity() {
    return this.integrity;
  }
}

File: compiler/src/main/kotlin/io/github/apl_cornell/viaduct/security/solver/LabelConstant.kt
package io.github.apl_cornell.viaduct.security.solver

import io.github.apl_cornell.viaduct.algebra.FreeDistributiveLattice
import io.github.apl_cornell.viaduct.algebra.solver.ConstantTerm
import io.github.apl_cornell.viaduct.security.Label
import io.github.apl_cornell.viaduct.security.Principal

/** Terms representing literal label constants. */
data class LabelConstant(val value: Label) : AtomicLabelTerm() {
    override fun getValue(solution: ConstraintSolution): Label {
        return value
    }

    override fun confidentiality(): LabelConstant {
        return LabelConstant(value.confidentiality())
    }

    override fun integrity(): LabelConstant {
        return LabelConstant(value.integrity())
    }

    override fun swap(): LabelConstant {
        return LabelConstant(value.swap())
    }

    override fun join(that: Label): LabelConstant {
        return LabelConstant(value.join(that))
    }

    override val confidentialityComponent: ConstantTerm<FreeDistributiveLattice<Principal>>
        get() = ConstantTerm(value.confidentialityComponent)

    override val integrityComponent: ConstantTerm<FreeDistributiveLattice<Principal>>
        get() = ConstantTerm(value.integrityComponent)
}
